### ISSUE TITLE
"Test connectivity" button in datasource form

### ISSUE THREAD

**Issue by AntoineThebaud** [perses/perses#1542]
### Proposal

Currently when creating/editing a datasource, you have no idea if the configuration you provided actually works. Thus it would be great to have a simple button to launch a dummy query using the provided config, to validate that the connectivity is OK.

**Comment by adrianSepiol** [comment]
I can work on it.

**Comment by jgbernalp** [comment]
@adrianSepiol thnx for your contribution. I was checking the PR sequence and I wonder if you will be open to take a more pragmatic approach. I think we can reuse the DatasourceClient interface, which already includes a healthCheck function:

```typescript
// @perses-dev/plugin-system
export interface DatasourceClient {
  kind?: string;
  healthCheck?: () => Promise<boolean>;
}
```

This will allow plugins to implement their health check in their datasource clients, some of them already have it. Then, in the HTTPSettingsEditor we can create a client with the current spec and call the function. This will allow us to make less assumptions on how to call a specific URL or which method to use by delegating that to plugins. Also we will avoid creating a new provider as all we need is the plugin client, which can be created from the spec. I think the client can be created lazily as soon as the button is pressed so it uses the current spec values. WDYT? Did I miss something that would avoid us to take this path?

**Comment by adrianSepiol** [comment]
> [@adrianSepiol](https://github.com/adrianSepiol) thnx for your contribution. I was checking the PR sequence and I wonder if you will be open to take a more pragmatic approach. I think we can reuse the DatasourceClient interface, which already includes a healthCheck function:
> 
> // @perses-dev/plugin-system
> export interface DatasourceClient {
>   kind?: string;
>   healthCheck?: () => Promise<boolean>;
> }
> 
> This will allow plugins to implement their health check in their datasource clients, some of them already have it. Then, in the HTTPSettingsEditor we can create a client with the current spec and call the function. This will allow us to make less assumptions on how to call a specific URL or which method to use by delegating that to plugins. Also we will avoid creating a new provider as all we need is the plugin client, which can be created from the spec. I think the client can be created lazily as soon as the button is pressed so it uses the current spec values. WDYT? Did I miss something that would avoid us to take this path?

@jgbernalp 
So I was thinking about this in first place but that would require some changes. Plugin can implement healthCheck like promethues does right now but the issue with current implementation is it won't work with unsaved datasource. The reason is for unsaved datasource we have to send spec itself with proxied call config to a POST endpoint (you can have a look at testProxyConnection in UnsavedDatasourceProvider). So to make it work in a way you describe I think we would need to pass something like "DatasourceHttpClient" or however we would end up calling it which would be interface with different implementation for saved and unsaved datasource and then use this in plugins instead of fetch. 

So to summarize how I would see this implementation: 

1. DatasourceHttpClient interface Generic HTTP client with methods like get<T>, post<T>, etc. Abstracts away transport details from plugin developers.
2. Two implementations
SavedDatasourceHttpClient — resolves URL/headers from saved datasource definition, routes through backend proxy normally.
UnsavedDatasourceHttpClient — wraps the intended call inside a POST body (spec + call to proxy), sent to a backend endpoint that executes it on behalf of the unsaved datasource.
3. Provider DatasourceHttpClientProvider selects the correct implementation based on saved/unsaved state, exposes it via a hook (e.g. useDatasourceHttpClient()).
4. Plugin adjustments Plugins drop direct fetch/fetchJson calls and consume DatasourceHttpClient exclusively. Shared fetch-wrapper boilerplate across plugins gets consolidated.
 
Benefits
- Plugin devs are decoupled from proxy mechanics.
- Reduces repeated fetch-wrapper code across plugins.

Challenges
- Backward compatibility: plugins using DatasourceHttpClient won't work with older Perses versions that don't provide it — needs a fallback strategy or versioning contract.
- large refactor touching many plugins.

If you feel like this is worth the effort I could got that way. Let me know

**Comment by jgbernalp** [comment]
This might be a lot of work. What if we treat all the datasources as unsaved for the "Test connection" button. It makes sense as you don't want to save changes that don't work. So in the `DatasourceEditorForm` we add a new component `TestDatasourceConnectionButton` in which we get the spec and the plugin healthcheck path, instead of a healthcheck function so we can create the right fetch request. Inside the button we use the `unsaved` proxy or direct connection using the healthcheck path. This does not require any new providers or changes across plugins other that defining the path for the health check. WDYT?

**Comment by adrianSepiol** [comment]
@jgbernalp This new provider allows us to access context in which editor is open (dashboard, project, global) which is needed to create proxy url but this part I think we could pass to the component as prop whenever it is used. What I struggle with is datasourceApi.buildProxyUrl which has PERSES_APP_CONFIG.api_prefix encapsulated in it. I'm not sure how to make it accessible in such way that we don't drill down through lots of components with props. That's why I came up with the provider.

**Comment by jgbernalp** [comment]
@adrianSepiol, as you said the current context can be passed as a prop. PERSES_APP_CONFIG is a global variable stored in the window object, and the buildProxyUrl already accounts for it so we don't need to bother for this case, just use the `buildProxyUrl` function, WDYT?

**Comment by adrianSepiol** [comment]
@jgbernalp buildProxyUrl is defined in app, DatasourceEditorForm is in plugin-system and is used in app and in dashboard. To avoid heavy props drilling in dashboard context we would need to move buildProxyUrl somewhere that plugin-system/dashboard can import from or expose wrapper function from DatasourceStoreProvider. What do you think?

**Comment by jgbernalp** [comment]
True, we were discussing this with @shahrokni, last week. We will create a new package inside `@perses-dev/shared` called `client` which has details about interacting with the perses backend, including this logic to build the proxy URL.

**Comment by adrianSepiol** [comment]
@jgbernalp I've updated the PRs for this issue. I got rid or the provider and simply call hook with project/dashboard to create correct context for test connection call.
